lib/logfmt.c

```diff
@@ -156,6 +162,113 @@ EXPORTED void logfmt_escape(struct buf *buf, const char *val)
     if (need_dquotes) buf_putc(buf, '\"');
 }
 
+static inline bool is_utf8_escaped_hex(uint32_t cp)
+{
+    switch (cp) {
+    case 0x000b: /* LINE TABULATION */
+    case 0x000c: /* FORM FEED */
+    case 0x0085: /* NEXT LINE */
+    case 0x2028: /* LINE SEPARATOR */
+    case 0x2029: /* PARAGRAPH SEPARATOR */
+        return true;
+    default:
+        return u_iscntrl(cp);
+    }
+}
+
+/* visible for testing; probably don't call this directly */
+EXPORTED void logfmt_escape_utf8(struct buf *buf, const char *utf8val)
+{
+    UText ut = UTEXT_INITIALIZER; /* XXX cache between calls? */
+    UErrorCode status = U_ZERO_ERROR;
+    bool need_dquotes;
+    uint32_t cp;
+
+    if (utf8val == NULL) {
+        buf_setcstr(buf, "~null~");
+        return;
+    }
+
+    if (!utf8val[0]) {
+        buf_setcstr(buf, "\"\"");
+        return;
+    }
+
+    /* utext_openUTF8:
+     * > Any invalid UTF-8 in the input will be handled in this way: a sequence
+     * > of bytes that has the form of a truncated, but otherwise valid, UTF-8
+     * > sequence will be replaced by a single unicode replacement character,
+     * > \uFFFD. Any other illegal bytes will each be replaced by a \uFFFD.
+     *
+     * The "replacement character" \uFFFD is �.  This character doesn't need to
+     * be escaped for logfmt, so we don't need (and can't have) explicit error
+     * handling for invalid utf8.
+     */
+    utext_openUTF8(&ut, utf8val, -1, &status);
+    if (U_FAILURE(status)) {
+        /* it's not clear what circumstances this might error in, but fall back
+         * to a bytestring escape anyway
+         */
+        syslog(LOG_DEBUG, "%s: utext_openUTF8() failed with status %s",
+                          __func__, u_errorName(status));
+        logfmt_escape_bytestring(buf, utf8val);
+        return;
+    }
+
+    need_dquotes = val_needs_dquotes(utf8val);
+
+    buf_reset(buf);
+    buf_ensure(buf, strlen(utf8val) + 2 * need_dquotes);
+    if (need_dquotes) buf_putc(buf, '"');
+
+    while ((uint32_t) U_SENTINEL != (cp = utext_next32(&ut))) {
+        uint8_t scratch[8] = {0};
+        unsigned i, len = 0;
+        bool is_error = false;
```

Comment:
nit: the argument types for `U8_APPEND` are `UBool` for `isError`. The 2nd and 3rd numeric arguments for buffer length and capacity are `int32_t`